src/WebSocket.cs

```diff
@@ -70,23 +71,81 @@ public async Task Connect(string? auth, string host, string nameOrAddress, Addre
             try
             {
                 await Ws.ConnectAsync(url, source.Token);
-                if (OnConnect != null) dispatchQueue.Enqueue(() => OnConnect());
+                if (Ws.State == WebSocketState.Open)
+                {
+                    if (OnConnect != null)
+                    {
+                        dispatchQueue.Enqueue(() => OnConnect());
+                    }
+                }
+                else
+                {
+                    if (OnConnectError != null)
+                    {
+                        dispatchQueue.Enqueue(() => OnConnectError(
+                            new Exception($"WebSocket connection failed. Current state: {Ws.State}")));
+                    }
+                    return;
+                }
+            }
+            catch (WebSocketException ex) when (ex.WebSocketErrorCode == WebSocketError.Success)
+            {
+                // How can we get here:
+                // - When you go to connect and the server isn't running (port closed) - target machine actively refused
+                // - 404 - No module with at that module address instead of 101 upgrade
+                // - 401? - When the identity received by SpacetimeDB wasn't signed by its signing key
+                // - 400 - When the auth is malformed
+                if (OnConnectError != null)
+                {
+                    // .net 6,7,8 has support for Ws.HttpStatusCode as long as you set
```

Comment:
Could we have a #ifdef here with support for .net 6+ ?

Reply:
This is a good idea, but I’d like to do it in a separate PR since the current PR is already an improvement for the customers we have